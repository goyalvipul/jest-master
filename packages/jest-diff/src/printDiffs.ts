/**
 * Copyright (c) Facebook, Inc. and its affiliates. All Rights Reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import chalk from 'chalk';

import {
  cleanupSemantic,
  DIFF_EQUAL,
  DIFF_DELETE,
  DIFF_INSERT,
  Diff,
} from './cleanupSemantic';
import diffStrings from './diffStrings';
import getAlignedDiffs from './getAlignedDiffs';
import {
  joinAlignedDiffsExpand,
  joinAlignedDiffsNoExpand,
} from './joinAlignedDiffs';
import {DiffOptions} from './types';

export const DIM_COLOR = chalk.dim;
export const EXPECTED_COLOR = chalk.green;
export const INVERTED_COLOR = chalk.inverse;
export const RECEIVED_COLOR = chalk.red;
const PATCH_COLOR = chalk.yellow;

// Given change op and array of diffs, return concatenated string:
// * include common strings
// * include change strings which have argument op (inverse highlight)
// * exclude change strings which have opposite op
export const getHighlightedString = (op: number, diffs: Array<Diff>): string =>
  diffs.reduce(
    (reduced: string, diff: Diff): string =>
      reduced +
      (diff[0] === DIFF_EQUAL
        ? diff[1]
        : diff[0] === op
        ? INVERTED_COLOR(diff[1])
        : ''),
    '',
  );

export const getExpectedString = (diffs: Array<Diff>): string =>
  getHighlightedString(DIFF_DELETE, diffs);

export const getReceivedString = (diffs: Array<Diff>): string =>
  getHighlightedString(DIFF_INSERT, diffs);

export const MULTILINE_REGEXP = /\n/;

const NEWLINE_SYMBOL = '\u{21B5}'; // downwards arrow with corner leftwards
const SPACE_SYMBOL = '\u{00B7}'; // middle dot

// Instead of inverse highlight which now implies a change,
// replace common spaces with middle dot at the end of the line.
const replaceSpacesAtEnd = (line: string): string =>
  line.replace(/\s+$/, spaces => SPACE_SYMBOL.repeat(spaces.length));

export const printDeleteLine = (line: string) =>
  EXPECTED_COLOR(line.length !== 0 ? '- ' + replaceSpacesAtEnd(line) : '-');

export const printInsertLine = (line: string) =>
  RECEIVED_COLOR(line.length !== 0 ? '+ ' + replaceSpacesAtEnd(line) : '+');

// Prevent visually ambiguous empty line as the first or the last.
export const printCommonLine = (line: string, isFirstOrLast: boolean = false) =>
  line.length !== 0
    ? DIM_COLOR('  ' + replaceSpacesAtEnd(line))
    : isFirstOrLast
    ? DIM_COLOR('  ' + NEWLINE_SYMBOL)
    : '';

export const computeStringDiffs = (expected: string, received: string) => {
  const isMultiline =
    MULTILINE_REGEXP.test(expected) || MULTILINE_REGEXP.test(received);

  // getAlignedDiffs assumes that a newline was appended to the strings.
  if (isMultiline) {
    expected += '\n';
    received += '\n';
  }

  const diffs = diffStrings(expected, received);
  cleanupSemantic(diffs); // impure function

  return {diffs, isMultiline};
};

export const hasCommonDiff = (diffs: Array<Diff>, isMultiline: boolean) => {
  if (isMultiline) {
    // Important: Ignore common newline that was appended to multiline strings!
    const iLast = diffs.length - 1;
    return diffs.some(
      (diff, i) => diff[0] === DIFF_EQUAL && (i !== iLast || diff[1] !== '\n'),
    );
  }

  return diffs.some(diff => diff[0] === DIFF_EQUAL);
};

export const printAnnotation = (options?: DiffOptions): string =>
  EXPECTED_COLOR('- ' + ((options && options.aAnnotation) || 'Expected')) +
  '\n' +
  RECEIVED_COLOR('+ ' + ((options && options.bAnnotation) || 'Received')) +
  '\n\n';

// In GNU diff format, indexes are one-based instead of zero-based.
export const createPatchMark = (
  aStart: number,
  aEnd: number,
  bStart: number,
  bEnd: number,
): string =>
  PATCH_COLOR(
    `@@ -${aStart + 1},${aEnd - aStart} +${bStart + 1},${bEnd - bStart} @@`,
  );

// Return formatted diff lines without labels.
export const printMultilineStringDiffs = (
  diffs: Array<Diff>,
  expand: boolean,
): string => {
  const lines = getAlignedDiffs(diffs);
  return expand
    ? joinAlignedDiffsExpand(lines)
    : joinAlignedDiffsNoExpand(lines);
};

const MAX_DIFF_STRING_LENGTH = 20000;

type StringDiffResult =
  | {isMultiline: true; annotatedDiff: string}
  | {isMultiline: false; a: string; b: string}
  | null;

// Print specific substring diff for strings only:
// * if strings are not equal
// * if neither string is empty
// * if neither string is too long
// * if there is a common string after semantic cleanup
export const getStringDiff = (
  expected: string,
  received: string,
  options?: DiffOptions,
): StringDiffResult => {
  if (
    expected === received ||
    expected.length === 0 ||
    received.length === 0 ||
    expected.length > MAX_DIFF_STRING_LENGTH ||
    received.length > MAX_DIFF_STRING_LENGTH
  ) {
    return null;
  }

  const {diffs, isMultiline} = computeStringDiffs(expected, received);

  if (!hasCommonDiff(diffs, isMultiline)) {
    return null;
  }

  return isMultiline
    ? {
        annotatedDiff:
          printAnnotation(options) +
          printMultilineStringDiffs(
            diffs,
            options === undefined || options.expand !== false,
          ),
        isMultiline,
      }
    : {
        a: getExpectedString(diffs),
        b: getReceivedString(diffs),
        isMultiline,
      };
};
