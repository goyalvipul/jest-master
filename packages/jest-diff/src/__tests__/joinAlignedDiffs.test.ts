/**
 * Copyright (c) Facebook, Inc. and its affiliates. All Rights Reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import {DIFF_DELETE, DIFF_EQUAL, DIFF_INSERT, Diff} from '../cleanupSemantic';
import {INVERTED_COLOR} from '../printDiffs';
import {
  joinAlignedDiffsExpand,
  joinAlignedDiffsNoExpand,
} from '../joinAlignedDiffs';

const diffsCommonStartEnd = [
  new Diff(DIFF_EQUAL, ''),
  new Diff(DIFF_EQUAL, 'common 2 preceding A'),
  new Diff(DIFF_EQUAL, 'common 1 preceding A'),
  new Diff(DIFF_DELETE, 'delete line'),
  new Diff(DIFF_DELETE, ['change ', INVERTED_COLOR('expect'), 'ed A'].join('')),
  new Diff(DIFF_INSERT, ['change ', INVERTED_COLOR('receiv'), 'ed A'].join('')),
  new Diff(DIFF_EQUAL, 'common 1 following A'),
  new Diff(DIFF_EQUAL, 'common 2 following A'),
  new Diff(DIFF_EQUAL, 'common 3 following A'),
  new Diff(DIFF_EQUAL, 'common 4 following A'),
  new Diff(DIFF_EQUAL, 'common 4 preceding B'),
  new Diff(DIFF_EQUAL, 'common 3 preceding B'),
  new Diff(DIFF_EQUAL, 'common 2 preceding B'),
  new Diff(DIFF_EQUAL, 'common 1 preceding B'),
  new Diff(DIFF_DELETE, ['change ', INVERTED_COLOR('expect'), 'ed B'].join('')),
  new Diff(DIFF_INSERT, ['change ', INVERTED_COLOR('receiv'), 'ed B'].join('')),
  new Diff(DIFF_INSERT, 'insert line'),
  new Diff(DIFF_EQUAL, 'common 1 following B'),
  new Diff(DIFF_EQUAL, 'common 2 following B'),
  new Diff(DIFF_EQUAL, 'common 3 between B and C'),
  new Diff(DIFF_EQUAL, 'common 2 preceding C'),
  new Diff(DIFF_EQUAL, 'common 1 preceding C'),
  new Diff(DIFF_DELETE, ['change ', INVERTED_COLOR('expect'), 'ed C'].join('')),
  new Diff(DIFF_INSERT, ['change ', INVERTED_COLOR('receiv'), 'ed C'].join('')),
  new Diff(DIFF_EQUAL, 'common 1 following C'),
  new Diff(DIFF_EQUAL, 'common 2 following C'),
  new Diff(DIFF_EQUAL, 'common 3 following C'),
  new Diff(DIFF_EQUAL, ''),
  new Diff(DIFF_EQUAL, 'common 5 following C'),
];

const diffsChangeStartEnd = [
  new Diff(DIFF_DELETE, 'delete'),
  new Diff(DIFF_EQUAL, 'common following delete'),
  new Diff(DIFF_EQUAL, 'common preceding insert'),
  new Diff(DIFF_INSERT, 'insert'),
];

describe('joinAlignedDiffsExpand', () => {
  test('first line is empty common', () => {
    expect(joinAlignedDiffsExpand(diffsCommonStartEnd)).toMatchSnapshot();
  });
});

describe('joinAlignedDiffsNoExpand', () => {
  test('patch 0 with context 1 and change at start and end', () => {
    expect(joinAlignedDiffsNoExpand(diffsChangeStartEnd, 1)).toMatchSnapshot();
  });

  test('patch 0 with context 5 and first line is empty common', () => {
    expect(joinAlignedDiffsNoExpand(diffsCommonStartEnd)).toMatchSnapshot();
  });

  test('patch 1 with context 4 and last line is empty common', () => {
    expect(joinAlignedDiffsNoExpand(diffsCommonStartEnd, 4)).toMatchSnapshot();
  });

  test('patch 2 with context 3', () => {
    expect(joinAlignedDiffsNoExpand(diffsCommonStartEnd, 3)).toMatchSnapshot();
  });

  test('patch 3 with context 2 and omit excess common at start', () => {
    expect(joinAlignedDiffsNoExpand(diffsCommonStartEnd, 2)).toMatchSnapshot();
  });
});
