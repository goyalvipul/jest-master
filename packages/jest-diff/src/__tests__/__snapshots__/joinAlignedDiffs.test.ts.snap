// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`joinAlignedDiffsExpand first line is empty common 1`] = `
"<dim>  ↵</>
<dim>  common 2 preceding A</>
<dim>  common 1 preceding A</>
<green>- delete line</>
<green>- change <inverse>expect</>ed A</>
<red>+ change <inverse>receiv</>ed A</>
<dim>  common 1 following A</>
<dim>  common 2 following A</>
<dim>  common 3 following A</>
<dim>  common 4 following A</>
<dim>  common 4 preceding B</>
<dim>  common 3 preceding B</>
<dim>  common 2 preceding B</>
<dim>  common 1 preceding B</>
<green>- change <inverse>expect</>ed B</>
<red>+ change <inverse>receiv</>ed B</>
<red>+ insert line</>
<dim>  common 1 following B</>
<dim>  common 2 following B</>
<dim>  common 3 between B and C</>
<dim>  common 2 preceding C</>
<dim>  common 1 preceding C</>
<green>- change <inverse>expect</>ed C</>
<red>+ change <inverse>receiv</>ed C</>
<dim>  common 1 following C</>
<dim>  common 2 following C</>
<dim>  common 3 following C</>

<dim>  common 5 following C</>"
`;

exports[`joinAlignedDiffsNoExpand patch 0 with context 1 and change at start and end 1`] = `
"<green>- delete</>
<dim>  common following delete</>
<dim>  common preceding insert</>
<red>+ insert</>"
`;

exports[`joinAlignedDiffsNoExpand patch 0 with context 5 and first line is empty common 1`] = `
"<dim>  ↵</>
<dim>  common 2 preceding A</>
<dim>  common 1 preceding A</>
<green>- delete line</>
<green>- change <inverse>expect</>ed A</>
<red>+ change <inverse>receiv</>ed A</>
<dim>  common 1 following A</>
<dim>  common 2 following A</>
<dim>  common 3 following A</>
<dim>  common 4 following A</>
<dim>  common 4 preceding B</>
<dim>  common 3 preceding B</>
<dim>  common 2 preceding B</>
<dim>  common 1 preceding B</>
<green>- change <inverse>expect</>ed B</>
<red>+ change <inverse>receiv</>ed B</>
<red>+ insert line</>
<dim>  common 1 following B</>
<dim>  common 2 following B</>
<dim>  common 3 between B and C</>
<dim>  common 2 preceding C</>
<dim>  common 1 preceding C</>
<green>- change <inverse>expect</>ed C</>
<red>+ change <inverse>receiv</>ed C</>
<dim>  common 1 following C</>
<dim>  common 2 following C</>
<dim>  common 3 following C</>

<dim>  common 5 following C</>"
`;

exports[`joinAlignedDiffsNoExpand patch 1 with context 4 and last line is empty common 1`] = `
"<yellow>@@ -1,24 +1,24 @@</>

<dim>  common 2 preceding A</>
<dim>  common 1 preceding A</>
<green>- delete line</>
<green>- change <inverse>expect</>ed A</>
<red>+ change <inverse>receiv</>ed A</>
<dim>  common 1 following A</>
<dim>  common 2 following A</>
<dim>  common 3 following A</>
<dim>  common 4 following A</>
<dim>  common 4 preceding B</>
<dim>  common 3 preceding B</>
<dim>  common 2 preceding B</>
<dim>  common 1 preceding B</>
<green>- change <inverse>expect</>ed B</>
<red>+ change <inverse>receiv</>ed B</>
<red>+ insert line</>
<dim>  common 1 following B</>
<dim>  common 2 following B</>
<dim>  common 3 between B and C</>
<dim>  common 2 preceding C</>
<dim>  common 1 preceding C</>
<green>- change <inverse>expect</>ed C</>
<red>+ change <inverse>receiv</>ed C</>
<dim>  common 1 following C</>
<dim>  common 2 following C</>
<dim>  common 3 following C</>
<dim>  ↵</>"
`;

exports[`joinAlignedDiffsNoExpand patch 2 with context 3 1`] = `
"<yellow>@@ -1,8 +1,7 @@</>

<dim>  common 2 preceding A</>
<dim>  common 1 preceding A</>
<green>- delete line</>
<green>- change <inverse>expect</>ed A</>
<red>+ change <inverse>receiv</>ed A</>
<dim>  common 1 following A</>
<dim>  common 2 following A</>
<dim>  common 3 following A</>
<yellow>@@ -11,13 +10,14 @@</>
<dim>  common 3 preceding B</>
<dim>  common 2 preceding B</>
<dim>  common 1 preceding B</>
<green>- change <inverse>expect</>ed B</>
<red>+ change <inverse>receiv</>ed B</>
<red>+ insert line</>
<dim>  common 1 following B</>
<dim>  common 2 following B</>
<dim>  common 3 between B and C</>
<dim>  common 2 preceding C</>
<dim>  common 1 preceding C</>
<green>- change <inverse>expect</>ed C</>
<red>+ change <inverse>receiv</>ed C</>
<dim>  common 1 following C</>
<dim>  common 2 following C</>
<dim>  common 3 following C</>"
`;

exports[`joinAlignedDiffsNoExpand patch 3 with context 2 and omit excess common at start 1`] = `
"<yellow>@@ -2,6 +2,5 @@</>
<dim>  common 2 preceding A</>
<dim>  common 1 preceding A</>
<green>- delete line</>
<green>- change <inverse>expect</>ed A</>
<red>+ change <inverse>receiv</>ed A</>
<dim>  common 1 following A</>
<dim>  common 2 following A</>
<yellow>@@ -12,5 +11,6 @@</>
<dim>  common 2 preceding B</>
<dim>  common 1 preceding B</>
<green>- change <inverse>expect</>ed B</>
<red>+ change <inverse>receiv</>ed B</>
<red>+ insert line</>
<dim>  common 1 following B</>
<dim>  common 2 following B</>
<yellow>@@ -18,5 +18,5 @@</>
<dim>  common 2 preceding C</>
<dim>  common 1 preceding C</>
<green>- change <inverse>expect</>ed C</>
<red>+ change <inverse>receiv</>ed C</>
<dim>  common 1 following C</>
<dim>  common 2 following C</>"
`;
